File: backend/back/src/main/java/com/back/domain/welfare/policy/entity/Policy.java
package com.back.domain.welfare.policy.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "policy")
@Getter
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Policy {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String plcyNo; // 정책번호
    private String plcyNm; // 정책명
    private String plcyKywdNm; // 정책키워드명

    @Lob
    private String plcyExplnCn; // 정책설명내용
    // 너무 긴 반환값을 가지는 칼럼에 한해 큰 객체로 저장하는 어노테이션 추가
    @Lob
    private String plcySprtCn; // 정책지원내용

    private String sprvsnInstCdNm; // 주관기관코드명(주관기관명)
    private String operInstCdNm; // 운영기관코드명(운영기관명)

    private String aplyPrdSeCd; // 신청기간구분코드(상시, 특정기간 등)

    private String bizPrdBgngYmd; // 사업기간시작일자
    private String bizPrdEndYmd; // 사업기간종료일자

    @Lob
    private String plcyAplyMthdCn; // 정책신청방법내용

    @Lob
    private String aplyUrlAddr; // 신청URL주소

    @Lob
    private String sbmsnDcmntCn; // 제출서류내용

    private String sprtTrgtMinAge; // 지원대상최소연령
    private String sprtTrgtMaxAge; // 지원대상최대연령
    private String sprtTrgtAgeLmtYn; // 지원대상연령제한여부

    private String mrgSttsCd; // 결혼상태코드
    private String earnCndSeCd; // 소득조건구분코드(무관, 연소득, 기타)
    private String earnMinAmt; // 소득최소금액
    private String earnMaxAmt; // 소득최대금액

    @Lob
    private String zipCd; // 정책거주지역코드

    private String jobCd; // 정책취업요건코드
    private String schoolCd; // 정책학력요건코드

    private String aplyYmd; // 신청기간
    private String sBizCd; // 정책특화요건코드

    // 원본 JSON (정책 1건 단위)
    @Lob
    @Column(columnDefinition = "TEXT")
    private String rawJson;
}

File: frontend/src/types/policy.ts
// Policy 관련 타입 정의
// 백엔드 API 계약 기반
// GET /api/v1/welfare/policy/search

// ===== 검색 요청 파라미터 =====
// PolicyElasticSearchRequestDto 기반
export interface PolicySearchRequest {
  zipCd?: string | null;
  schoolCode?: string | null;
  jobCode?: string | null;
  keyword?: string | null;
  age?: number | null;
  earn?: number | null;
  regionCode?: string | null;
  marriageStatus?: string | null;
  keywords?: string[] | null;
  from: number;
  size: number;
}

// ===== 검색 응답 =====
// PolicyDocument 기반
export interface PolicyDocument {
  policyId: number | null;
  plcyNo: string | null;
  plcyNm: string | null;

  // 나이
  minAge: number | null;
  maxAge: number | null;
  ageLimited: boolean | null;

  // 소득
  earnCondition: string | null;
  earnMin: number | null;
  earnMax: number | null;

  // 대상 조건
  regionCode: string | null;
  jobCode: string | null;
  schoolCode: string | null;
  marriageStatus: string | null;

  // 태그 / 분류
  keywords: string[] | null;
  specialBizCode: string | null;

  // 검색용 텍스트
  description: string | null;
}

// ===== 북마크 응답 =====
// GET /api/v1/member/bookmark/welfare-bookmarks
// BookmarkPolicyResponseDto 기반
export interface BookmarkPolicyResponse {
  code: number;
  message: string;
  policies: Policy[] | null;
}

// ===== 북마크 토글 응답 =====
// POST /api/v1/member/bookmark/welfare-bookmarks/{policyId}
// BookmarkUpdateResponseDto 기반
export interface BookmarkUpdateResponse {
  code: number;
  message: string;
}

// ===== 신청 관련 =====
// GET /api/v1/member/policy-aply/welfare-applications
// Application 엔티티 그대로 직렬화 (Policy, Member 포함)
export interface ApplicationItem {
  id: number;
  policy: Policy;
  applicant: ApplicationMember;
  createdAt: string; // ISO 8601
  modifiedAt: string; // ISO 8601
}

// Application 내 Member (엔티티 전체 직렬화)
export interface ApplicationMember {
  id: number;
  name: string;
  email: string;
}

// POST /api/v1/member/policy-aply/welfare-application/{policyId}
// AddApplicationResponseDto 기반
export interface AddApplicationResponse {
  status: number;
  message: string;
}

// PUT /api/v1/member/policy-aply/welfare-application/{id}
// DeleteApplicationResponseDto 기반
export interface DeleteApplicationResponse {
  code: number;
  message: string;
}

// Policy 엔티티 기반
export interface Policy {
  id: number | null;
  plcyNo: string | null;
  plcyNm: string | null;
  plcyKywdNm: string | null;
  plcyExplnCn: string | null;
  plcySprtCn: string | null;
  sprvsnInstCdNm: string | null;
  operInstCdNm: string | null;
  aplyPrdSeCd: string | null;
  bizPrdBgngYmd: string | null;
  bizPrdEndYmd: string | null;
  plcyAplyMthdCn: string | null;
  aplyUrlAddr: string | null;
  sbmsnDcmntCn: string | null;
  sprtTrgtMinAge: string | null;
  sprtTrgtMaxAge: string | null;
  sprtTrgtAgeLmtYn: string | null;
  mrgSttsCd: string | null;
  earnCndSeCd: string | null;
  earnMinAmt: string | null;
  earnMaxAmt: string | null;
  zipCd: string | null;
  jobCd: string | null;
  schoolCd: string | null;
  aplyYmd: string | null;
  sbizCd: string | null;
  rawJson: string | null;
}
